python/ray/serve/batching.py

```diff
@@ -445,33 +449,90 @@ def _validate_batch_wait_timeout_s(batch_wait_timeout_s):
         )
 
 
+SelfType = TypeVar("SelfType", contravariant=True)
 T = TypeVar("T")
 R = TypeVar("R")
-F = TypeVar("F", Callable[[List[T]], List[R]], Callable[[Any, List[T]], List[R]])
-G = TypeVar("G", bound=Callable[[T], R])
 
 
-# Normal decorator use case (called with no arguments).
-@overload
-def batch(func: F) -> G:
-    pass
+class _SyncBatchingMethod(Protocol, Generic[SelfType, T, R]):
+    def __call__(self, self_: SelfType, __batch: List[T], /) -> List[R]:
+        ...
 
 
-# "Decorator factory" use case (called with arguments).
-@overload
+class _AsyncBatchingMethod(Protocol, Generic[SelfType, T, R]):
+    async def __call__(_, self: SelfType, __batch: List[T], /) -> List[R]:
+        ...
+
+
+@overload  # Sync function for `batch` called WITHOUT arguments
+def batch(_sync_func: Callable[[List[T]], List[R]], /) -> Callable[[T], R]:
+    ...
+
+
+@overload  # Async function for `batch` called WITHOUT arguments
+def batch(
+    _async_func: Callable[[List[T]], Coroutine[Any, Any, List[R]]], /
+) -> Callable[[T], Coroutine[Any, Any, R]]:
+    ...
+
+
+@overload  # Sync method for `batch` called WITHOUT arguments
+def batch(
+    _sync_meth: _SyncBatchingMethod[SelfType, T, R], /
+) -> Callable[[SelfType, T], R]:
+    ...
+
+
+@overload  # Async method for `batch` called WITHOUT arguments
 def batch(
+    _async_meth: _AsyncBatchingMethod[SelfType, T, R], /
+) -> Callable[[SelfType, T], Coroutine[Any, Any, R]]:
+    ...
+
+
+@overload  # `batch` called WITH arguments
+def batch(
+    _: Literal[None] = None,
+    /,
     max_batch_size: int = 10,
     batch_wait_timeout_s: float = 0.0,
-) -> Callable[[F], G]:
-    pass
+) -> "_BatchDecorator":
+    ...
+
+
+class _BatchDecorator(Protocol):
+    """Descibes behaviour of decorator produced by calling `batch` with arguments"""
+
+    @overload  # Sync function
+    def __call__(self, _sync_func: Callable[[List[T]], List[R]], /) -> Callable[[T], R]:
+        ...
+
+    @overload  # Async function
+    def __call__(
+        self, _async_func: Callable[[List[T]], Coroutine[Any, Any, List[R]]], /
+    ) -> Callable[[T], Coroutine[Any, Any, R]]:
+        ...
+
+    @overload  # Sync method
+    def __call__(
+        self, _sync_meth: _SyncBatchingMethod[SelfType, T, R], /
+    ) -> Callable[[SelfType, T], R]:
+        ...
+
+    @overload  # Async method
+    def __call__(
+        self, _async_meth: _AsyncBatchingMethod[SelfType, T, R], /
+    ) -> Callable[[SelfType, T], Coroutine[Any, Any, R]]:
+        ...
 
 
 @PublicAPI(stability="stable")
 def batch(
     _func: Optional[Callable] = None,
+    /,
```

Comment:
Why add the slash here?

Reply:
The slash here signifies that the first argument is positional, and the rest must be provided as keywords, making it more explicit that the function should be called either as:

```python
@serve.batch # decorated_method is the positional, first argument
def decorated_method()
```

OR 
```python
@serve.batch(max_batch_size=?, batch_wait_timeout_s=?) # max_batch_size and batch_wait_timeout_s must be provided as key words as they are otherwise passed in the first positional argument which is reserved for the function.
def decorated_method()
```

Reply:
Is this different than using `*`? (that's what we use in other parts of the codebase)

Reply:
Yes, slightly different. 
The slash means that `_func` must be passed as a positional parameter, and `max_batch_size` and `batch_wait_timeout_s` must be are keyword-only.

If we use a *, we could still pass `_func` as a keyword. 

I went with the / because the _func is always passed as a positional when decorating.

Reply:
I found this table [here](https://geekpython.in/slash-and-asterisk-in-function#:~:text=A%20slash%20(%20%2F%20)%20in%20the,be%20treated%20as%20keyword%2Donly.):
## Parameter Types

| **Left-Side Parameters**       | **Symbol** | **Right-Side Parameters**         |
|--------------------------------|------------|-----------------------------------|
| Positional-only arguments      | /          | Either positional or keyword arguments |
| Either positional or keyword arguments | *          | Keyword-only arguments             |

Reply:
Oh very cool! In all the thousands of lines of Python code I've read and written, I've never encountered this lol